src/System.Buffers.Primitives/System/Buffers/ReadOnlyBuffer.cs

```diff
@@ -389,5 +397,35 @@ public void Reset()
                 ThrowHelper.ThrowNotSupportedException();
             }
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private ReadOnlyBuffer<T> SliceImpl(Position begin, Position end)
+        {
+            return new ReadOnlyBuffer<T>(
+                begin.Segment,
+                begin.Index | (Start.Index & ~IndexBitMask),
+                end.Segment,
+                end.Index | (End.Index & ~IndexBitMask)
+            );
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private  BufferType GetBufferType()
+        {
+            return (BufferType)((((uint)Start.Index >> 30) & 2) | (uint)End.Index >> 31);
```

Comment:
This is really ugly. I'm sure there is a better way.

Reply:
Use the low bits instead?
```csharp
return (BufferType)((Start.Index & 1) << 1) | End.Index & 1);
```
Then to get value its 
```csharp
private static int GetIndex(int index)
{
    return index >> 1;
}
```
Add flagging; shift 1 + or
```csharp
BufferStart = new Position(first, (0 << 1) | MemoryListStartFlag);
BufferEnd = new Position(segment, (segment.Memory.Length << 1) | MemoryListEndFlag);
```

Reply:
Interesting idea. I guess we don't really care which bit to use.

`return index >> 1;` wouldn't shift cause problems if we set sign bit? It would start filling with ones instead of zeros.

Reply:
Store it as a `uint`?

Reply:
`>> 1` is less ugly than `& 0x7FFFFFFF` e.g. ` & IndexBitMask`
And
`& 1` is less ugly than `& 0x80000000` e.g. ` & ~IndexBitMask`